File: c/part2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

/* ANSI colors (same palette as part1) */
#define RST  "\033[0m"
#define BOLD "\033[1m"
#define DIM  "\033[2m"
#define RED  "\033[91m"
#define GRN  "\033[92m"
#define YLW  "\033[93m"
#define MAG  "\033[95m"
#define CYN  "\033[96m"
#define GRY  "\033[90m"

#define CL  "  " DIM "│" RST "  "

static void drain_line(void) {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

static void code_open(void) {
    printf(DIM "  ╭─ " RST BOLD RED "c" RST DIM " ──────────────────────────────────────────────────────\n");
    printf("  │\n" RST);
}

static void code_close(void) {
    printf(DIM "  │\n");
    printf("  ╰──────────────────────────────────────────────────────────\n\n" RST);
}

static void section_result(void) {
    printf(BOLD CYN "\n  ━━ Result " RST DIM "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n\n" RST);
}

static void section_explain(void) {
    printf(BOLD MAG "\n  ━━ What happened " RST DIM "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n\n" RST);
}

static void press_enter(const char *question) {
    printf(BOLD YLW "\n  ╔═ ? ════════════════════════════════════════════════════════╗\n" RST);
    printf(BOLD YLW "  ║" RST "  %s\n", question);
    printf(BOLD YLW "  ║\n" RST);
    printf(BOLD YLW "  ║" RST "  " DIM "Press Enter to see what actually happens..." RST "\n");
    printf(BOLD YLW "  ╚════════════════════════════════════════════════════════════╝\n" RST);
    fflush(stdout);
    drain_line();
    printf("\n");
}

/* ============================================================
 * Demo 1: Dangling pointer — return address of local variable
 * ============================================================ */

/* Route through a global so GCC can't replace the return with NULL.
   GCC detects direct 'return &local' and zeroes it as a hardening
   measure; going via a global defeats that analysis. */
static volatile void *g_leak;

__attribute__((noinline))
static int *get_dangling(int val) {
    int local = val;
    g_leak = &local;  /* capture address before return */
    return (int *)(uintptr_t)g_leak;  /* return stale address */
}

static void demo_dangling(void) {
    code_open();
    printf(CL YLW "int" RST " *get_dangling() {\n");
    printf(CL "    " YLW "int" RST " local = " MAG "42" RST ";\n");
    printf(CL "    " GRY "// stack frame exists here\n" RST);
    printf(CL "    " YLW "return" RST " &local;  " RED "// ← stack frame destroyed on return" RST "\n");
    printf(CL "}\n");
    printf(DIM "  │\n" RST);
    printf(CL YLW "int" RST " *p = get_dangling();\n");
    printf(CL GRY "// p now points into a destroyed stack frame\n" RST);
    printf(CL CYN "printf" RST "(" GRN "\"%%d\\n\"" RST ", *p);         " GRY "// ← what will this print?" RST "\n");
    printf(CL "get_dangling(" RED "0xDEAD" RST ");        " GRY "// ← reuse the same stack frame" RST "\n");
    printf(CL CYN "printf" RST "(" GRN "\"%%d\\n\"" RST ", *p);         " GRY "// ← what about now?" RST "\n");
    code_close();

    press_enter("p points to a local that no longer exists. What will *p be after the frame is reused?");
    section_result();

    int *p = get_dangling(42);
    printf("  [after return]  p = %p\n", (void *)p);
    printf("  *p immediately  = " YLW BOLD "%d" RST "  (frame not yet reused)\n", *p);
    get_dangling(0xDEAD);  /* second call: same frame, different value */
    printf("  *p after reuse  = " RED BOLD "0x%X" RST "  ← first call's 'local' corrupted!\n", (unsigned int)*p);

    section_explain();
    printf("  The pointer is " RED BOLD "dangling" RST ": it refers to memory the function\n");
    printf("  no longer owns. The first read returned " YLW "42" RST " only because\n");
    printf("  nothing had reused that address yet.\n\n");
    printf("  A second call to the same function " RED "reuses the identical frame\n" RST);
    printf("  layout — same address, now containing the new call's " YLW "0xDEAD" RST ".\n");
    printf("  p still points there: " RED BOLD "same pointer, different data, no warning" RST ".\n\n");
    printf("  This is why returning " RED "&local_variable" RST " is always wrong:\n");
    printf("    " RED "•" RST " Correct by accident today\n");
    printf("    " RED "•" RST " Silently broken when the stack layout changes\n");
    printf("    " RED "•" RST " Impossible to detect without tooling\n\n");
    printf("  Run " CYN "'make run2-asan'" RST " to see AddressSanitizer catch this.\n");
}

/* ============================================================
 * Demo 2: Use-after-free
 * ============================================================ */
static void demo_use_after_free(void) {
    code_open();
    printf(CL YLW "int" RST " *p = " CYN "malloc" RST "(" YLW "sizeof" RST "(" YLW "int" RST "));\n");
    printf(CL "*p = " MAG "100" RST ";\n");
    printf(CL CYN "free" RST "(p);               " GRY "// memory returned to allocator" RST "\n");
    printf(DIM "  │\n" RST);
    printf(CL CYN "printf" RST "(" GRN "\"%%d\\n\"" RST ", *p); " GRY "// ← read freed memory" RST "\n");
    printf(CL "*p = " RED "999" RST ";              " RED "// ← write freed memory" RST "\n");
    code_close();

    printf("  " DIM "The allocator now owns those bytes. We still have the address.\n\n" RST);

    press_enter("After free(), what will *p read as? Will the write to *p cause an immediate crash?");
    section_result();

    int *p = malloc(sizeof(int));
    *p = 100;
    printf("  Before free: *p = " GRN BOLD "%d" RST "\n", *p);
    free(p);
    printf("  After free:  *p = " RED BOLD "%d" RST "  (value may still linger...)\n", *p);

    /* Trigger allocator activity to corrupt the freed block, then read again */
    void *noise = malloc(sizeof(int));
    free(noise);
    printf("  After alloc/free cycle: *p = " RED BOLD "%d" RST "  (allocator touched it)\n", *p);

    *p = 999;  /* write to freed memory */
    printf("  After *p=999: *p = " RED BOLD "%d" RST "  (wrote to freed heap)\n", *p);

    section_explain();
    printf("  The read after free often returns the old value because the\n");
    printf("  allocator hasn't zeroed the memory. This makes the bug " RED BOLD "invisible\n" RST);
    printf("  in normal testing but " RED BOLD "exploitable" RST " — freed heap blocks contain\n");
    printf("  allocator bookkeeping metadata that an attacker can read.\n\n");
    printf("  The write to freed memory silently corrupts the heap. The\n");
    printf("  crash (if any) happens " RED BOLD "later" RST ", at an unrelated allocation,\n");
    printf("  making the root cause nearly " RED "impossible to trace" RST " without tooling.\n\n");
    printf("  Run " CYN "'make run2-asan'" RST " to see AddressSanitizer pinpoint this.\n");
}

/* ============================================================
 * Demo 3: Double-free — runs in a child process to survive the crash
 * ============================================================ */
static void demo_double_free(void) {
    code_open();
    printf(CL YLW "int" RST " *p = " CYN "malloc" RST "(" YLW "sizeof" RST "(" YLW "int" RST "));\n");
    printf(CL "*p = " MAG "42" RST ";\n");
    printf(CL CYN "free" RST "(p);  " GRY "// correct — first free" RST "\n");
    printf(CL CYN "free" RST "(p);  " RED "// ← second free — double free!" RST "\n");
    code_close();

    printf("  " DIM "Running in a child process so this demo can't kill the shell.\n" RST);

    press_enter("What happens when you free() the same pointer twice?");
    section_result();

    fflush(stdout);
    pid_t pid = fork();
    if (pid == 0) {
        /* child */
        int *p = malloc(sizeof(int));
        *p = 42;
        free(p);
        free(p);  /* double-free: undefined behavior */
        _exit(0); /* should not reach here */
    }

    int status;
    waitpid(pid, &status, 0);

    if (WIFSIGNALED(status)) {
        int sig = WTERMSIG(status);
        const char *signame = (sig == SIGABRT)  ? "SIGABRT" :
                              (sig == SIGSEGV)  ? "SIGSEGV" :
                              (sig == SIGBUS)   ? "SIGBUS"  : "unknown";
        printf("  Child process killed by " RED BOLD "signal %d (%s)" RST "\n\n", sig, signame);
        printf("  " DIM "The crash came from inside malloc's own bookkeeping code,\n");
        printf("  not from the second free() call site.\n" RST);
    } else if (WIFEXITED(status)) {
        printf("  Child exited normally (exit %d) — allocator accepted the double-free.\n",
               WEXITSTATUS(status));
        printf("  " YLW BOLD "Silent heap corruption" RST " — harder to detect than a crash.\n");
    }

    section_explain();
    printf("  Heap allocators track free blocks in a linked list or bitmap.\n");
    printf("  A double-free corrupts that internal structure.\n\n");
    printf("  Outcomes (all possible, all undefined behavior):\n");
    printf("    " RED "•" RST " Crash with SIGABRT from the allocator consistency check\n");
    printf("    " RED "•" RST " Crash with SIGSEGV when the corrupt list is next walked\n");
    printf("    " RED "•" RST " Silent corruption that triggers a crash in unrelated code later\n");
    printf("    " RED "•" RST " Exploitable: attacker shapes heap to make double-free\n");
    printf("      return the same block twice — " RED BOLD "two pointers to one allocation" RST "\n\n");
    printf("  Run " CYN "'make run2-asan'" RST " to see AddressSanitizer catch this cleanly.\n");
}

/* ============================================================
 * Demo 4: Null pointer — forgotten NULL check
 * ============================================================ */
static int *find_value(int *arr, size_t len, int target) {
    for (size_t i = 0; i < len; i++)
        if (arr[i] == target) return &arr[i];
    return NULL;  /* not found */
}

static void demo_null_deref(void) {
    code_open();
    printf(CL YLW "int" RST " *find_value(" YLW "int" RST " *arr, " YLW "size_t" RST " len, " YLW "int" RST " target) {\n");
    printf(CL "    " YLW "for" RST " (...) " YLW "if" RST " (arr[i] == target) " YLW "return" RST " &arr[i];\n");
    printf(CL "    " YLW "return" RST " " RED "NULL" RST ";   " GRY "// not found" RST "\n");
    printf(CL "}\n");
    printf(DIM "  │\n" RST);
    printf(CL YLW "int" RST " arr[] = {" MAG "1" RST "," MAG "2" RST "," MAG "3" RST "};\n");
    printf(CL YLW "int" RST " *p = find_value(arr, " MAG "3" RST ", " RED "99" RST "); " GRY "// 99 not in array" RST "\n");
    printf(CL GRY "// forgot to check: if (p == NULL) ...\n" RST);
    printf(CL CYN "printf" RST "(" GRN "\"%%d\\n\"" RST ", *p);  " RED "// ← dereference NULL" RST "\n");
    code_close();

    printf("  " DIM "99 is not in the array. find_value returns NULL.\n" RST);

    press_enter("find_value returns NULL when not found. We dereference it without checking. What happens?");
    section_result();

    int arr[] = {1, 2, 3};
    int *found = find_value(arr, 3, 2);
    printf("  find_value(arr, 3, 2)  = " GRN BOLD "%d" RST "  (found)\n", *found);

    fflush(stdout);
    pid_t pid = fork();
    if (pid == 0) {
        int *p = find_value(arr, 3, 99);
        printf("  find_value(arr, 3, 99) returned pointer %p\n", (void *)p);
        printf("  Dereferencing NULL...\n");
        fflush(stdout);
        printf("  *p = %d\n", *p);  /* crashes here */
        _exit(0);
    }

    int status;
    waitpid(pid, &status, 0);
    if (WIFSIGNALED(status)) {
        int sig = WTERMSIG(status);
        printf("  Dereference NULL → " RED BOLD "signal %d (SIGSEGV)" RST "\n", sig);
    }

    section_explain();
    printf("  NULL is address 0x0. The OS maps no memory there.\n");
    printf("  Dereferencing it always causes a " RED BOLD "segmentation fault" RST ".\n\n");
    printf("  The silent problem: " BOLD "nothing forces you to check the return value" RST ".\n");
    printf("  In C, NULL is just a convention — the type of find_value is\n");
    printf("  " YLW "int *" RST ", which could be a valid pointer or NULL. There is no\n");
    printf("  difference in the type system.\n\n");
    printf("  Every pointer in C is implicitly " RED "nullable" RST ". The compiler\n");
    printf("  will not warn you if you forget the NULL check.\n");
    printf("  Result: NULL dereferences are one of the most common C crashes.\n");
}

/* ============================================================ */

static void print_menu(void) {
    printf(RED BOLD "\n  ┌──────────────────────────────────────┐\n");
    printf("  │      C Pointer Lifetime Horror       │\n");
    printf("  └──────────────────────────────────────┘\n" RST);
    printf(DIM "  Part 2: dangling pointers, use-after-free, double-free, null.\n\n" RST);
    printf("  " CYN "1)" RST " Dangling pointer (return &local)\n");
    printf("  " CYN "2)" RST " Use-after-free\n");
    printf("  " CYN "3)" RST " Double-free (runs in a subprocess)\n");
    printf("  " CYN "4)" RST " Null pointer — forgotten NULL check\n");
    printf("  " DIM "q) Quit\n" RST);
    printf("\n  " BOLD "> " RST);
    fflush(stdout);
}

int main(void) {
    int ch;
    while (1) {
        print_menu();
        ch = getchar();
        if (ch == EOF) break;
        drain_line();
        printf("\n");

        switch (ch) {
            case '1': demo_dangling();       break;
            case '2': demo_use_after_free(); break;
            case '3': demo_double_free();    break;
            case '4': demo_null_deref();     break;
            case 'q': case 'Q':
                printf(DIM "\n  Exiting. Leaks not included.\n\n" RST);
                return 0;
            default:
                printf(RED "  Unknown option '%c'.\n" RST, ch);
        }
    }
    return 0;
}

File: typescript/src/part2.ts
import * as readline from "readline";

const RST  = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM  = "\x1b[2m";
const RED  = "\x1b[91m";
const GRN  = "\x1b[92m";
const YLW  = "\x1b[93m";
const BLU  = "\x1b[94m";
const MAG  = "\x1b[95m";
const CYN  = "\x1b[96m";
const GRY  = "\x1b[90m";

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let rlClosed = false;
rl.on("close", () => { rlClosed = true; });

function ask(question: string): Promise<string> {
  return new Promise((resolve) => {
    if (rlClosed) { resolve("q"); return; }
    const onClose = () => resolve("q");
    rl.once("close", onClose);
    rl.question(question, (answer) => {
      rl.removeListener("close", onClose);
      resolve(answer);
    });
  });
}

const CL = `  ${DIM}│${RST}  `;

function codeOpen(lang: string, color: string): void {
  console.log(`${DIM}  ╭─ ${RST}${BOLD}${color}${lang}${RST} ${DIM}────────────────────────────────────────────────────────`);
  console.log(`  │${RST}`);
}
const codeOpenC  = () => codeOpen("c",  RED);
const codeOpenTs = () => codeOpen("ts", BLU);

function codeClose(): void {
  console.log(`${DIM}  │`);
  console.log(`  ╰──────────────────────────────────────────────────────────${RST}`);
  console.log();
}

function sectionResult(): void {
  console.log(`${BOLD}${CYN}\n  ━━ Result ${RST}${DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${RST}\n`);
}

function sectionExplain(): void {
  console.log(`${BOLD}${MAG}\n  ━━ What happened ${RST}${DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${RST}\n`);
}

async function pressEnter(question: string): Promise<void> {
  console.log(`${BOLD}${YLW}\n  ╔═ ? ════════════════════════════════════════════════════════╗${RST}`);
  console.log(`${BOLD}${YLW}  ║${RST}  ${question}`);
  console.log(`${BOLD}${YLW}  ║${RST}`);
  console.log(`${BOLD}${YLW}  ║${RST}  ${DIM}Press Enter to see what actually happens...${RST}`);
  console.log(`${BOLD}${YLW}  ╚════════════════════════════════════════════════════════════╝${RST}`);
  await ask("");
  console.log();
}

// ============================================================
// Demo 1: GC prevents dangling — references keep objects alive
// ============================================================
async function demoNoDangling(): Promise<void> {
  console.log(`  In C, returning a pointer to a local variable is dangling UB:\n`);

  codeOpenC();
  console.log(`${CL}${YLW}int${RST} *bad_get() {`);
  console.log(`${CL}    ${YLW}int${RST} local = ${MAG}42${RST};`);
  console.log(`${CL}    ${RED}return &local;${RST}  ${GRY}// UB: stack frame gone on return${RST}`);
  console.log(`${CL}}`);
  codeClose();

  console.log(`  In TypeScript, returning an object reference is always safe:\n`);

  codeOpenTs();
  console.log(`${CL}${YLW}function${RST} getRef(): ${GRN}{ value: number }${RST} {`);
  console.log(`${CL}    ${YLW}const${RST} obj = { value: ${MAG}42${RST} };`);
  console.log(`${CL}    ${YLW}return${RST} obj;  ${GRY}// passes reference — GC sees caller holds it${RST}`);
  console.log(`${CL}}`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${CL}${YLW}const${RST} ref1 = getRef();`);
  console.log(`${CL}${YLW}const${RST} ref2 = getRef();`);
  console.log(`${CL}console.log(ref1.value, ref2.value);  ${GRY}// always valid${RST}`);
  codeClose();

  await pressEnter("No stack to destroy — what lifetime does a JS object have?");
  sectionResult();

  function getRef() { return { value: 42 }; }

  const ref1 = getRef();
  const ref2 = getRef();
  console.log(`  ref1.value = ${GRN}${BOLD}${ref1.value}${RST}  (alive: ref1 holds it)`);
  console.log(`  ref2.value = ${GRN}${BOLD}${ref2.value}${RST}  (alive: ref2 holds it)\n`);

  // Demonstrate multiple references to same object
  const obj = { value: 99 };
  const a = obj;
  const b = obj;
  a.value = 100;
  console.log(`  Shared reference: a and b both point to the same object`);
  console.log(`  a.value = ${GRN}${BOLD}${a.value}${RST}`);
  console.log(`  b.value = ${GRN}${BOLD}${b.value}${RST}  (same object, not a copy)`);

  sectionExplain();
  console.log(`  JavaScript objects live on the ${BOLD}heap${RST}, managed by a garbage collector.`);
  console.log(`  An object is kept alive as long as ${BOLD}any reference to it exists${RST}.\n`);
  console.log(`  When getRef() returns, the local variable ${GRN}obj${RST} is gone — but`);
  console.log(`  the ${BOLD}reference${RST} was returned to the caller, so the GC sees the`);
  console.log(`  object is still reachable. It will not be collected.\n`);
  console.log(`  There is ${BOLD}no stack${RST} in the C sense. Objects are not tied to`);
  console.log(`  function call frames. The concept of a dangling pointer ${GRN}does not exist${RST}.`);
}

// ============================================================
// Demo 2: No use-after-free — GC owns the free timing
// ============================================================
async function demoNoUseAfterFree(): Promise<void> {
  console.log(`  In C, free() returns memory to the allocator immediately:\n`);

  codeOpenC();
  console.log(`${CL}${YLW}int${RST} *p = ${CYN}malloc${RST}(${YLW}sizeof${RST}(${YLW}int${RST}));`);
  console.log(`${CL}*p = ${MAG}42${RST};`);
  console.log(`${CL}${CYN}free${RST}(p);           ${GRY}// heap freed now${RST}`);
  console.log(`${CL}${YLW}int${RST} x = *p;        ${RED}// ← use-after-free: UB${RST}`);
  codeClose();

  console.log(`  In TypeScript, you cannot free memory — the GC decides when:\n`);

  codeOpenTs();
  console.log(`${CL}${YLW}let${RST} owned: ${GRN}{ value: number }${RST} | ${RED}null${RST} = { value: ${MAG}42${RST} };`);
  console.log(`${CL}${YLW}const${RST} shared = owned;    ${GRY}// second reference${RST}`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${CL}owned = ${RED}null${RST};           ${GRY}// drop our reference${RST}`);
  console.log(`${CL}${GRY}// GC will NOT collect: 'shared' still holds a reference${RST}`);
  console.log(`${CL}console.log(shared.value); ${GRN}// always safe${RST}`);
  codeClose();

  await pressEnter("Setting a variable to null — does that free the object? What if another variable holds it?");
  sectionResult();

  let owned: { value: number } | null = { value: 42 };
  const shared = owned;

  console.log(`  Before: owned.value = ${GRN}${BOLD}${owned.value}${RST}, shared.value = ${GRN}${BOLD}${shared.value}${RST}`);

  owned = null;
  console.log(`  After owned = null:`);
  console.log(`    owned  = ${YLW}${owned}${RST}`);
  console.log(`    shared.value = ${GRN}${BOLD}${shared.value}${RST}  (object still alive via shared)`);

  // Show that WeakRef loses access after GC
  console.log();
  console.log(`  WeakRef — holds a reference the GC can ignore:\n`);
  let strong: object | null = { secret: "still here" };
  const weak = new WeakRef(strong);
  console.log(`  weak.deref() before release = ${GRN}${BOLD}${JSON.stringify(weak.deref())}${RST}`);
  strong = null;
  console.log(`  strong = null (only weak reference remains)`);
  console.log(`  ${DIM}(GC may collect this — deref() could return undefined after GC runs)${RST}`);
  console.log(`  weak.deref() right now      = ${YLW}${BOLD}${JSON.stringify(weak.deref())}${RST}  ${DIM}(not yet collected)${RST}`);

  sectionExplain();
  console.log(`  ${BOLD}You cannot free memory in JavaScript${RST}. Setting a variable to ${RED}null${RST}`);
  console.log(`  just removes ${BOLD}your${RST} reference. The GC tracks all references and`);
  console.log(`  only frees the object when ${BOLD}none remain${RST}.\n`);
  console.log(`  ${GRN}WeakRef${RST} is the only way to hold a reference the GC can ignore.`);
  console.log(`  After the strong reference is gone, ${YLW}weak.deref()${RST} may return`);
  console.log(`  ${RED}undefined${RST} at any point after the next GC cycle.\n`);
  console.log(`  The tradeoff: you cannot ${BOLD}control when${RST} memory is freed.`);
  console.log(`  For most apps this is fine. For tight memory budgets (games,`);
  console.log(`  embedded, high-throughput servers) GC pauses can be a problem.`);
  console.log(`  This is one reason Rust exists — ${GRN}deterministic deallocation${RST}`);
  console.log(`  without a GC.`);
}

// ============================================================
// Demo 3: No double-free — the concept does not exist
// ============================================================
async function demoNoDoubleFree(): Promise<void> {
  console.log(`  In C, you can call free() on the same pointer twice:\n`);

  codeOpenC();
  console.log(`${CL}${YLW}int${RST} *p = ${CYN}malloc${RST}(${YLW}sizeof${RST}(${YLW}int${RST}));`);
  console.log(`${CL}${CYN}free${RST}(p);  ${GRY}// correct${RST}`);
  console.log(`${CL}${CYN}free${RST}(p);  ${RED}// double-free: heap corruption / SIGABRT${RST}`);
  codeClose();

  console.log(`  In TypeScript, there is no free() — GC owns all deallocation:\n`);

  codeOpenTs();
  console.log(`${CL}${YLW}let${RST} a: ${GRN}object${RST} | ${RED}null${RST} = { value: ${MAG}42${RST} };`);
  console.log(`${CL}${YLW}let${RST} b = a;        ${GRY}// second reference to same object${RST}`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${CL}a = ${RED}null${RST};          ${GRY}// remove one reference${RST}`);
  console.log(`${CL}b = ${RED}null${RST};          ${GRY}// remove last reference${RST}`);
  console.log(`${CL}${GRY}// GC now eligible to collect — no double-free possible${RST}`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${CL}a = ${RED}null${RST};          ${GRY}// nulling null: perfectly fine${RST}`);
  console.log(`${CL}b = ${RED}null${RST};          ${GRY}// same — no-op, no error${RST}`);
  codeClose();

  await pressEnter("There's no free(). How does memory ever actually get reclaimed?");
  sectionResult();

  let a: { value: number; refs?: string } | null = { value: 42, refs: "a,b" };
  let b: typeof a = a;

  console.log(`  [a and b point to same object]  a===b: ${GRN}${BOLD}${a === b}${RST}`);

  a = null;
  console.log(`  After a=null: b.value = ${GRN}${BOLD}${b?.value}${RST}  (still alive via b)`);

  b = null;
  console.log(`  After b=null: no references remain`);
  console.log(`  ${DIM}GC will collect the object at its next cycle — we have no`);
  console.log(`  handle, no address, no way to access or double-free it${RST}\n`);

  console.log(`  Attempting to "double-null" is a harmless no-op:`);
  a = null;
  b = null;
  console.log(`  a = null again: ${GRN}${BOLD}ok${RST}  (a = ${a}, b = ${b})`);

  sectionExplain();
  console.log(`  The GC maintains a ${BOLD}reference count${RST} (or reachability graph).`);
  console.log(`  An object is freed exactly ${BOLD}once${RST}: when the count reaches zero.\n`);
  console.log(`  You cannot trigger a double-free because:`);
  console.log(`    ${GRN}•${RST} There is no ${RED}free()${RST} to call`);
  console.log(`    ${GRN}•${RST} Setting a variable to ${RED}null${RST} just decrements the count`);
  console.log(`    ${GRN}•${RST} The GC calls the internal free exactly once\n`);
  console.log(`  The cost: ${YLW}non-deterministic timing${RST}. You cannot predict exactly`);
  console.log(`  when memory is reclaimed. For resource-constrained systems`);
  console.log(`  or real-time requirements, this is why manual/ownership-based`);
  console.log(`  memory management (C, Rust) is still used.`);
}

// ============================================================
// Demo 4: undefined as null — optional chaining / null safety
// ============================================================
function findValue(arr: number[], target: number): number | undefined {
  return arr.find(x => x === target);
}

async function demoNullSafety(): Promise<void> {
  console.log(`  In C, find_value returns NULL — type is just int *, you must check:\n`);

  codeOpenC();
  console.log(`${CL}${YLW}int${RST} *p = find_value(arr, n, ${RED}99${RST}); ${GRY}// might be NULL${RST}`);
  console.log(`${CL}${GRY}// C type: int * — same whether valid or NULL${RST}`);
  console.log(`${CL}${CYN}printf${RST}(${GRN}"%%d\\n"${RST}, *p);  ${RED}// crash if NULL — nothing warned you${RST}`);
  codeClose();

  console.log(`  In TypeScript the return type encodes the absent case:\n`);

  codeOpenTs();
  console.log(`${CL}${YLW}function${RST} findValue(arr: ${GRN}number[]${RST}, t: ${YLW}number${RST}): ${GRN}number | undefined${RST} {`);
  console.log(`${CL}    ${YLW}return${RST} arr.${CYN}find${RST}(x => x === t);`);
  console.log(`${CL}}`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${CL}${YLW}const${RST} v = findValue(arr, ${RED}99${RST});`);
  console.log(`${CL}v${RED}.toFixed(2)${RST};     ${RED}// TS error: v is number | undefined${RST}`);
  console.log(`${CL}v${GRN}?.toFixed(2)${RST};    ${GRN}// safe: optional chain, returns undefined${RST}`);
  console.log(`${CL}(v ${GRN}?? 0${RST}).toFixed(${MAG}2${RST}); ${GRN}// safe: fallback to 0${RST}`);
  codeClose();

  await pressEnter("TypeScript knows findValue might return undefined. What does that mean at runtime?");
  sectionResult();

  const arr = [1, 2, 3, 4, 5];

  for (const target of [3, 99]) {
    const v = findValue(arr, target);
    if (v !== undefined) {
      console.log(`  findValue(arr, ${String(target).padStart(2)}) = ${GRN}${BOLD}${v}${RST}  (found)`);
    } else {
      console.log(`  findValue(arr, ${String(target).padStart(2)}) = ${RED}${BOLD}undefined${RST}  (not found)`);
    }
  }

  console.log();
  const v = findValue(arr, 99);  // undefined
  console.log(`  v?.toFixed(2)       = ${YLW}${BOLD}${v?.toFixed(2)}${RST}   (optional chain short-circuits)`);
  console.log(`  (v ?? 0).toFixed(2) = ${GRN}${BOLD}${(v ?? 0).toFixed(2)}${RST}  (nullish coalescing provides default)`);
  console.log(`  v ?? "not found"    = ${GRN}${BOLD}"${v ?? "not found"}"${RST}`);

  sectionExplain();
  console.log(`  TypeScript's ${RED}null${RST} / ${RED}undefined${RST} in a union type makes the absent`);
  console.log(`  case ${BOLD}visible in the type system${RST}. The compiler rejects code that`);
  console.log(`  calls methods on ${GRN}number | undefined${RST} without narrowing first.\n`);
  console.log(`  Three patterns for safe access:`);
  console.log(`    ${GRN}v?.method()${RST}    — optional chaining: returns undefined if v is nil`);
  console.log(`    ${GRN}v ?? fallback${RST}  — nullish coalescing: substitute a default`);
  console.log(`    ${GRN}if (v !== undefined)${RST} — type narrowing: v is ${YLW}number${RST} inside\n`);
  console.log(`  Enable ${YLW}strictNullChecks${RST} in tsconfig (included in ${YLW}strict${RST}) to get`);
  console.log(`  this protection. Without it, ${RED}null${RST} and ${RED}undefined${RST} are assignable`);
  console.log(`  to every type — the TS equivalent of C's unguarded NULL pointer.`);
}

function printMenu(): void {
  console.log(`${BLU}${BOLD}\n  ┌──────────────────────────────────────┐`);
  console.log(`  │    TypeScript / GC Memory Safety     │`);
  console.log(`  └──────────────────────────────────────┘${RST}`);
  console.log(`${DIM}  Part 2: GC lifetimes, no free(), WeakRef, null safety.\n${RST}`);
  console.log(`  ${CYN}1)${RST} No dangling — GC keeps objects alive`);
  console.log(`  ${CYN}2)${RST} No use-after-free — GC owns the free timing`);
  console.log(`  ${CYN}3)${RST} No double-free — concept doesn't exist`);
  console.log(`  ${CYN}4)${RST} Null safety — undefined in the type system`);
  console.log(`  ${DIM}q) Quit${RST}`);
}

async function main(): Promise<void> {
  while (true) {
    printMenu();
    const choice = await ask(`\n  ${BOLD}> ${RST}`);
    console.log();

    switch (choice.trim()) {
      case "1": await demoNoDangling();    break;
      case "2": await demoNoUseAfterFree(); break;
      case "3": await demoNoDoubleFree();  break;
      case "4": await demoNullSafety();    break;
      case "q": case "Q":
        console.log(`${DIM}\n  Exiting. Memory reclaimed whenever the GC gets around to it.\n${RST}`);
        rl.close();
        return;
      default:
        console.log(`${RED}  Unknown option.${RST}`);
    }
  }
}

main();
